# Credit Card Kaggle Anamoly Detection


## Context

It is important that credit card companies are able to recognize fraudulent credit card transactions so that customers are not charged for items that they did not purchase.

## Content

The datasets contains transactions made by credit cards in September 2013 by european cardholders. This dataset presents transactions that occurred in two days, where we have 492 frauds out of 284,807 transactions. The dataset is highly unbalanced, the positive class (frauds) account for 0.172% of all transactions.

It contains only numerical input variables which are the result of a PCA transformation. Unfortunately, due to confidentiality issues, we cannot provide the original features and more background information about the data. Features V1, V2, ... V28 are the principal components obtained with PCA, the only features which have not been transformed with PCA are 'Time' and 'Amount'. Feature 'Time' contains the seconds elapsed between each transaction and the first transaction in the dataset. The feature 'Amount' is the transaction Amount, this feature can be used for example-dependant cost-senstive learning. Feature 'Class' is the response variable and it takes value 1 in case of fraud and 0 otherwise.

## Inspiration

dentify fraudulent credit card transactions.

Given the class imbalance ratio, we recommend measuring the accuracy using the Area Under the Precision-Recall Curve (AUPRC). Confusion matrix accuracy is not meaningful for unbalanced classification.

## Acknowledgements

The dataset has been collected and analysed during a research collaboration of Worldline and the Machine Learning Group (http://mlg.ulb.ac.be) of ULB (Université Libre de Bruxelles) on big data mining and fraud detection. More details on current and past projects on related topics are available on https://www.researchgate.net/project/Fraud-detection-5 and the page of the DefeatFraud project

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('creditcard.csv')
data.head()


Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
Time      284807 non-null float64
V1        284807 non-null float64
V2        284807 non-null float64
V3        284807 non-null float64
V4        284807 non-null float64
V5        284807 non-null float64
V6        284807 non-null float64
V7        284807 non-null float64
V8        284807 non-null float64
V9        284807 non-null float64
V10       284807 non-null float64
V11       284807 non-null float64
V12       284807 non-null float64
V13       284807 non-null float64
V14       284807 non-null float64
V15       284807 non-null float64
V16       284807 non-null float64
V17       284807 non-null float64
V18       284807 non-null float64
V19       284807 non-null float64
V20       284807 non-null float64
V21       284807 non-null float64
V22       284807 non-null float64
V23       284807 non-null float64
V24       284807 non-null float64
V25       284807 non-null float64
V26  

## Exploratory Data Analysis

In [4]:
data.isnull().values.any()

False

In [5]:
data.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

we can see it's imbalanced dataset.

In [6]:

## Get the Fraud and the normal dataset 

fraud = data[data['Class']==1]

normal = data[data['Class']==0]

In [7]:
print(fraud.shape,normal.shape)

(492, 31) (284315, 31)


In [8]:
## We need to analyze more amount of information from the transaction data
#How different are the amount of money used in different transaction classes?
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [9]:
normal.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [10]:
## Take some sample of the data

data1= data.sample(frac = 0.1,random_state=1)

data1.shape

(28481, 31)

In [11]:
data.shape

(284807, 31)

In [12]:
#Determine the number of fraud and valid transactions in the dataset

Fraud = data1[data1['Class']==1]

Valid = data1[data1['Class']==0]

outlier_fraction = len(Fraud)/float(len(Valid))

In [13]:
print(outlier_fraction)

print("Fraud Cases : {}".format(len(Fraud)))

print("Valid Cases : {}".format(len(Valid)))

0.0017234102419808666
Fraud Cases : 49
Valid Cases : 28432


In [14]:
data1.corr()

Time        V1        V2        V3        V4        V5        V6  \
Time    1.000000  0.126475 -0.001584 -0.413547 -0.104527  0.182205 -0.060483   
V1      0.126475  1.000000  0.048796  0.015452 -0.010592  0.019888  0.006417   
V2     -0.001584  0.048796  1.000000  0.027270 -0.022539  0.009666 -0.004411   
V3     -0.413547  0.015452  0.027270  1.000000 -0.005423  0.013997 -0.006903   
V4     -0.104527 -0.010592 -0.022539 -0.005423  1.000000 -0.003708  0.002029   
V5      0.182205  0.019888  0.009666  0.013997 -0.003708  1.000000 -0.016656   
V6     -0.060483  0.006417 -0.004411 -0.006903  0.002029 -0.016656  1.000000   
V7      0.078924 -0.020583 -0.013456 -0.024640  0.004432 -0.037463  0.006923   
V8     -0.040474 -0.003013  0.015662 -0.025529  0.011659 -0.013263  0.003695   
V9     -0.008428  0.001658  0.003456  0.002525 -0.004395 -0.008506 -0.002762   
V10     0.035939  0.010686  0.017218 -0.006955 -0.000669  0.011446 -0.003120   
V11    -0.237613  0.007177 -0.002982  0.000836  0.007657  0.006286  0.001582   
V12     0.126985 -0.012290  0.000646 -0.008236  0.005942 -0.011393 -0.000219   
V13    -0.069353 -0.018849  0.001655  0.004030 -0.011637 -0.002728 -0.001591   
V14    -0.093264 -0.001905 -0.006617 -0.016702  0.003485  0.001548 -0.007828   
V15    -0.182255 -0.012884  0.002046  0.004421 -0.000575 -0.006731  0.006407   
V16     0.007392 -0.015569 -0.003062 -0.012780 -0.004783 -0.014823  0.009237   
V17    -0.074555 -0.009644  0.003567 -0.017722  0.012553 -0.013090  0.009201   
V18     0.083959 -0.011822 -0.005922  0.001056  0.001852  0.000902 -0.002719   
V19     0.019469  0.003860  0.011950  0.020282  0.001759  0.001468 -0.005205   
V20    -0.050967 -0.017883 -0.054467  0.003068  0.015348  0.005350 -0.008991   
V21     0.041323 -0.016415 -0.020127 -0.006083 -0.004423  0.002288  0.004490   
V22     0.150603  0.014896  0.021923  0.014177 -0.011251  0.022065 -0.003705   
V23     0.047941  0.049447  0.047591  0.042603 -0.017682  0.064703 -0.036726   
V24    -0.020018 -0.003709 -0.011386 -0.001883  0.001829 -0.007184  0.001428   
V25    -0.229491  0.014055  0.011838  0.005975 -0.009692  0.006493 -0.015012   
V26    -0.048131  0.007203  0.005366  0.006869  0.004087  0.000048  0.009938   
V27    -0.005541 -0.011545 -0.009611 -0.017094  0.024489 -0.027934 -0.004811   
V28    -0.004339  0.085035  0.084873  0.029973 -0.024554  0.010991 -0.009772   
Amount -0.026969 -0.262703 -0.556401 -0.225099  0.111692 -0.397437  0.213007   
Class  -0.005087 -0.079820  0.069598 -0.160051  0.122631 -0.073519 -0.035085   

              V7        V8        V9  ...       V21       V22       V23  \
Time    0.078924 -0.040474 -0.008428  ...  0.041323  0.150603  0.047941   
V1     -0.020583 -0.003013  0.001658  ... -0.016415  0.014896  0.049447   
V2     -0.013456  0.015662  0.003456  ... -0.020127  0.021923  0.047591   
V3     -0.024640 -0.025529  0.002525  ... -0.006083  0.014177  0.042603   
V4      0.004432  0.011659 -0.004395  ... -0.004423 -0.011251 -0.017682   
V5     -0.037463 -0.013263 -0.008506  ...  0.002288  0.022065  0.064703   
V6      0.006923  0.003695 -0.002762  ...  0.004490 -0.003705 -0.036726   
V7      1.000000 -0.028291 -0.005510  ...  0.007012 -0.013871 -0.055242   
V8     -0.028291  1.000000 -0.018645  ... -0.005651 -0.004195  0.030092   
V9     -0.005510 -0.018645  1.000000  ...  0.009462 -0.002297  0.002360   
V10    -0.035149 -0.017995 -0.021718  ...  0.001434  0.006397  0.010754   
V11     0.000863 -0.002562  0.000587  ...  0.009261 -0.000061 -0.004784   
V12    -0.008740 -0.009387  0.002794  ... -0.016599 -0.002456 -0.000431   
V13     0.002445  0.012011  0.001733  ...  0.002637  0.004413 -0.017930   
V14    -0.001300 -0.001876  0.004310  ... -0.007459  0.001944  0.005999   
V15    -0.003969 -0.000946 -0.014436  ...  0.000210  0.002279 -0.013669   
V16    -0.010928 -0.000048  0.000286  ... -0.007622  0.004838 -0.026575   
V17    -0.017525 -0.020343 -0.007033  ... -0.015777  0.005118  0.013149   
V18    -0.015221 -0.013191

In [17]:
#Create independent and Dependent Features
columns = data1.columns.tolist()
# Filter the columns to remove data we do not want 
columns = [c for c in columns if c not in ["Class"]]
# Store the variable we are predicting 
target = "Class"
# Define a random state 
state = np.random.RandomState(42)
X = data1[columns]
Y = data1[target]
X_outliers = state.uniform(low=0, high=1, size=(X.shape[0], X.shape[1]))
# Print the shapes of X & Y
print(X.shape)
print(Y.shape)
print(X_outliers)

(28481, 30)
(28481,)
[[0.37454012 0.95071431 0.73199394 ... 0.51423444 0.59241457 0.04645041]
 [0.60754485 0.17052412 0.06505159 ... 0.19598286 0.04522729 0.32533033]
 [0.38867729 0.27134903 0.82873751 ... 0.63755747 0.88721274 0.47221493]
 ...
 [0.9759548  0.00835751 0.39654879 ... 0.27124459 0.49869169 0.0681891 ]
 [0.36655003 0.30495987 0.25223322 ... 0.42485128 0.20312083 0.42905276]
 [0.97759176 0.38214378 0.07343188 ... 0.04355135 0.6165196  0.17670918]]


In [16]:
#different method
X = data1.iloc[:,:-1]
Y = data.iloc[:,-1]
print(X.shape)
print(Y.shape)

(28481, 30)
(284807,)


## Model Prediction


Now it is time to start building the model .The types of algorithms we are going to use to try to do anomaly detection on this dataset are as follows


## Isolation Forest Algorithm 

One of the newest techniques to detect anomalies is called Isolation Forests. The algorithm is based on the fact that anomalies are data points that are few and different. As a result of these properties, anomalies are susceptible to a mechanism called isolation.

This method is highly useful and is fundamentally different from all existing methods. It introduces the use of isolation as a more effective and efficient means to detect anomalies than the commonly used basic distance and density measures. Moreover, this method is an algorithm with a low linear time complexity and a small memory requirement. It builds a good performing model with a small number of trees using small sub-samples of fixed size, regardless of the size of a data set.

Typical machine learning methods tend to work better when the patterns they try to learn are balanced, meaning the same amount of good and bad behaviors are present in the dataset.

How Isolation Forests Work The Isolation Forest algorithm isolates observations by randomly selecting a feature and then randomly selecting a split value between the maximum and minimum values of the selected feature. The logic argument goes: isolating anomaly observations is easier because only a few conditions are needed to separate those cases from the normal observations. On the other hand, isolating normal observations require more conditions. Therefore, an anomaly score can be calculated as the number of conditions required to separate a given observation.

The way that the algorithm constructs the separation is by first creating isolation trees, or random decision trees. Then, the score is calculated as the path length to isolate the observation.

## Local Outlier Factor(LOF) Algorithm

The LOF algorithm is an unsupervised outlier detection method which computes the local density deviation of a given data point with respect to its neighbors. It considers as outlier samples that have a substantially lower density than their neighbors.

The number of neighbors considered, (parameter n_neighbors) is typically chosen 1) greater than the minimum number of objects a cluster has to contain, so that other objects can be local outliers relative to this cluster, and 2) smaller than the maximum number of close by objects that can potentially be local outliers. In practice, such informations are generally not available, and taking n_neighbors=20 appears to work well in general.

In [20]:
from sklearn.metrics import classification_report,accuracy_score
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn.svm import OneClassSVM

In [21]:
##Define the outlier detection methods

classifiers = {
    "Isolation Forest":IsolationForest(n_estimators=100, max_samples=len(X), 
                                       contamination=outlier_fraction,random_state=state, verbose=0),
    "Local Outlier Factor":LocalOutlierFactor(n_neighbors=20, algorithm='auto', 
                                              leaf_size=30, metric='minkowski',
                                              p=2, metric_params=None, contamination=outlier_fraction),
    "Support Vector Machine":OneClassSVM(kernel='rbf', degree=3, gamma=0.1,nu=0.05, 
                                         max_iter=-1, random_state=state)
   
}

In [22]:
n_outliers = len(Fraud)
for i, (clf_name,clf) in enumerate(classifiers.items()):
    #Fit the data and tag outliers
    if clf_name == "Local Outlier Factor":
        y_pred = clf.fit_predict(X)
        scores_prediction = clf.negative_outlier_factor_
    elif clf_name == "Support Vector Machine":
        clf.fit(X)
        y_pred = clf.predict(X)
    else:    
        clf.fit(X)
        scores_prediction = clf.decision_function(X)
        y_pred = clf.predict(X)
    #Reshape the prediction values to 0 for Valid transactions , 1 for Fraud transactions
    y_pred[y_pred == 1] = 0
    y_pred[y_pred == -1] = 1
    n_errors = (y_pred != Y).sum()
    # Run Classification Metrics
    print("{}: {}".format(clf_name,n_errors))
    print("Accuracy Score :")
    print(accuracy_score(Y,y_pred))
    print("Classification Report :")
    print(classification_report(Y,y_pred))

C:\Users\venka\Anaconda3\lib\site-packages\sklearn\ensemble\iforest.py:247: FutureWarning: behaviour="old" is deprecated and will be removed in version 0.22. Please use behaviour="new", which makes the decision_function change to match other anomaly detection algorithm API.
  FutureWarning)
C:\Users\venka\Anaconda3\lib\site-packages\sklearn\ensemble\iforest.py:415: DeprecationWarning: threshold_ attribute is deprecated in 0.20 and will be removed in 0.22.
  " be removed in 0.22.", DeprecationWarning)


Isolation Forest: 73
Accuracy Score :
0.9974368877497279
Classification Report :
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     28432
           1       0.26      0.27      0.26        49

    accuracy                           1.00     28481
   macro avg       0.63      0.63      0.63     28481
weighted avg       1.00      1.00      1.00     28481

Local Outlier Factor: 97
Accuracy Score :
0.9965942207085425
Classification Report :
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     28432
           1       0.02      0.02      0.02        49

    accuracy                           1.00     28481
   macro avg       0.51      0.51      0.51     28481
weighted avg       1.00      1.00      1.00     28481



C:\Users\venka\Anaconda3\lib\site-packages\sklearn\svm\classes.py:1194: DeprecationWarning: The random_state parameter is deprecated and will be removed in version 0.22.
  " be removed in version 0.22.", DeprecationWarning)


Support Vector Machine: 8516
Accuracy Score :
0.7009936448860644
Classification Report :
              precision    recall  f1-score   support

           0       1.00      0.70      0.82     28432
           1       0.00      0.37      0.00        49

    accuracy                           0.70     28481
   macro avg       0.50      0.53      0.41     28481
weighted avg       1.00      0.70      0.82     28481



## Observations 

1.Isolation Forest detected 73 errors versus Local Outlier Factor detecting 97 errors vs. SVM detecting 8516 errors
2.Isolation Forest has a 99.74% more accurate than LOF of 99.65% and SVM of 70.09
3.When comparing error precision & recall for 3 models , the Isolation Forest performed much better than the LOF as we can see    that the detection of fraud cases is around 27 % versus LOF detection rate of just 2 % and SVM of 0%.
4.So overall Isolation Forest Method performed much better in determining the fraud cases which is around 30%.
5.We can also improve on this accuracy by increasing the sample size or use deep learning algorithms however at the cost of     computational expense.We can also use complex anomaly detection models to get better accuracy in determining more fraudulent cases